sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,608 @@
+        ++_ck_blocks_header_offset;
+        if (_ck_blocks_header_offset == 32u) {
+            _ck_blocks_header_offset = 0u;
+        }
+    }
+    std::optional<uint32_t> get_ck_block_value_length() const {
+        return _ck_column_value_fix_lengths.front();
+    }
+    bool is_block_empty() const {
+        return (_ck_blocks_header & (uint64_t(1) << (2 * _ck_blocks_header_offset))) != 0;
+    }
+    bool is_block_null() const {
+        return (_ck_blocks_header & (uint64_t(1) << (2 * _ck_blocks_header_offset + 1))) != 0;
+    }
+    bool should_read_block_header() const {
+        return _ck_blocks_header_offset == 0u;
+    }
+
+public:
+    bool non_consuming() const {
+        return false;
+    }
+    void verify_end_state() const {
+        if (!_finished) {
+            throw std::runtime_error("row_body_skipping_context - no more data but parsing is incomplete");
+        }
+    }
+    bool end_of_partition() const {
+        return _end_of_partition;
+    }
+    uint64_t prev_len() {
+        return _prev_unfiltered_size;
+    }
+    std::optional<tombstone_reversing_info> current_tombstone_reversing_info() {
+        // std::nullopt if the last consumed unfiltered was not a tombstone
+        return _current_tombstone_reversing_info;
+    }
+    data_consumer::processing_result process_state(temporary_buffer<char>& data) {
+        _processing_data = &data;
+        auto ret = _gen.generate();
+        return ret;
+    }
+private:
+    uint64_t current_position() {
+        return position() - _processing_data->size();
+    }
+    processing_result_generator do_process_state() {
+        while (true) {
+            _finished = false;
+            co_yield read_8(*_processing_data);
+            auto flags = unfiltered_flags_m(_u8);
+            _current_tombstone_reversing_info.reset();
+            if (flags.is_end_of_partition()) {
+                _end_of_partition = true;
+                _finished = true;
+                co_yield data_consumer::proceed::no;
+                break;
+            } else if (flags.is_range_tombstone()) {
+                _current_tombstone_reversing_info.emplace();
+                _current_tombstone_reversing_info->kind_offset = current_position();
+                co_yield read_8(*_processing_data);
+                _current_tombstone_reversing_info->range_tombstone_kind = bound_kind_m(_u8);
+                co_yield read_16(*_processing_data);
+                _ck_size = _u16;
+                if (_ck_size != 0) {
+                    _reading_range_tombstone_ck = true;
+                }
+            } else {
+                if (flags.has_extended_flags()) {
+                    // skip flags
+                    co_yield read_8(*_processing_data);
+                    // `row_body_skipping_context` should not be constructed on static rows
+                    auto extended_flags = unfiltered_extended_flags_m(_u8);
+                    assert(!extended_flags.is_static());
```

Comment:
on_internal_error() is more appropriate, we don't want to bring the cluster down on poisoned read

Reply:
We need a version of `on_internal_error` which can be used like an `assert` (without a separate `if`):
```
assert_internal(condition, logger, error_message)
```
which is equivalent to
```
if (!condition) on_internal_error(logger, error_message)
```
it would probably have to be a macro so we don't eagerly evaluate the `error_message` expression.